course-content/34_machine_learning.qmd

```diff
@@ -0,0 +1,244 @@
+
+Each of the predictions is based on a discrete probability. The `predict_proba()` method returns the class probabilities for each data point.  The number of probabilities for each row is equal to the number of categories in target variable (2 in our case).
+
+```{python}
+y_predict_proba =  reg.predict_proba(X_test) # computes a probability per case
+y_predict_proba[:5] # first 5 predictions in test dataset
+```
+
+We can see for each case a pair of probabilities, one for "0" and the other for "1". This detail helps us understand how the decisions with the `predict()` method were made. 
+
+
+## Evaluating
+
+Ok. We have our predictions... but are they good? Of course, what "good" means needs a bit of thought. There are many performance measures in data science and there are considerations based on the use cases themselves. How we would evaluate performace of an algorithm that attempts to diagnose serious disease would be different than if we were evaluating one that identified whether an email was spam or not. 
+
+There are a number of standard evaluation tools that are commonly used, however, and we will introduce a few here. 
+
+One of the most common and useful peformance scores is `accuracy_score()`, and it is fairly easy to interpret. It reflects that percentage of cases where the prediction of the model matches exactly the true labels of the sample.
+
+```{python}
+from sklearn.metrics import accuracy_score
+print(accuracy_score(y_test,y_predict))
+```
+
+In our case, it shows that 68% of the models guesses are correct. A standard comparison is to refer to the percentage of the most frequently occuring response in the training set. In our example, we designed that to be 50%. And since 68% > 50%, the model is doing something right!
+
+Another common performance measuring tool is the so-called "confusion matrix". 
+
+```{python}
+from sklearn.metrics import confusion_matrix
+print(confusion_matrix(y_test, y_predict)) 
+```
+
+
+<center>
+    <img src="images/machine-learning/confusion_matrix.PNG" width=70% style="margin:auto" border=1/>
+    <p style="text-align: center">
+    Four quadrants of the confusion matrix
+    </p>
+</center>
+
+The confusion matrix shows how many guesses the model gets "right", that is, the actual is true when we guess true (TP - True Positive) and false when we guess false (TN - True Negative). The remaining two quadrants show the errors. Type I error is the false positive error (FP - False Positive), where the model guesses true but reality is false. Type II error is where where the model guesses false but the reality is true (FN - False Negative).
+<br>
+<br>
+Another widely used report is the classification report, which shows precision and recall, which are two widely reported metrics in machine learning. 
+
+```{python}
+from sklearn.metrics import classification_report
+print(classification_report(y_test, y_predict))
+```
+
+**Precision** = True Positives / (True Positives + False Positives) <br>
```

Comment:
I find these are easier to wrap my head around with a little example/ english definition, but up to you. I like the cast a wide/narrow net analogy.

Reply:
Good idea. Net analogy is good.  Don't want to wade too deeply into these measures, as its more about the coding, but point very well taken. These are easy terms to forget from point 1 to point 2, as they are laden with cognitive baggage from other contexts.

Reply:
Yeah I agree, keep it simple!